### PR TITLE
Fix #5524.  Use finfo.max instead of np.inf

### LINKED ISSUE #5524: Change in colorbar extensions
In #5307, where I am lowering our testing tolerance down to zero, I am discovering a few recent changes that slipped through because the tolerance on our tests was too high.

In particular 8928311cc833ffa386344445709bcaa8e4b95243 introduced the following change in `test_axes:test_contour_colorbar`.

Current baseline image:

contour_colorbar-expected

Current behavior after 8928311cc833ffa386344445709bcaa8e4b95243:

contour_colorbar

The test code:

```
@image_comparison(baseline_images=['contour_colorbar'])
def test_contour_colorbar():
    x, y, z = contour_dat()

    fig = plt.figure()
    ax = fig.add_subplot(111)
    cs = ax.contourf(x, y, z, levels=np.arange(-1.8, 1.801, 0.2),
                     cmap=plt.get_cmap('RdBu'),
                     vmin=-0.6,
                     vmax=0.6,
                     extend='both')
    cs1 = ax.contour(x, y, z, levels=np.arange(-2.2, -0.599, 0.2),
                     colors=['y'],
                     linestyles='solid',
                     linewidths=2)
    cs2 = ax.contour(x, y, z, levels=np.arange(0.6, 2.2, 0.2),
                     colors=['c'],
                     linewidths=2)
    cbar = fig.colorbar(cs, ax=ax)
    cbar.add_lines(cs1)
    cbar.add_lines(cs2, erase=False)
```

@efiring: Do you know if this change was intentional?

### PR DESCRIPTION
**Pull request by mdboom**
Cc: @efiring 

Based on suggested fix in #5524.

### CODE REVIEW

**Comment by efiring** [comment]
I was thinking about the finfo values, but couldn't remember the name `finfo`.  I'm a little uneasy about it, though.  These values will be run through the norm, after which they will be checked against the 0-1 range.  The norming operation on these max values can generate overflow, yielding inf, leaving us right back where we started.  Although it is not as pretty, I think the practical solution may be to use a large value that still leaves scope for being normed without overflow.  If we used 1e200, for example, how often do you think that would cause trouble for a user?

**Comment by mdboom** [comment]
Stepping back, the problem seems to have been introduced when non-finite values were masked.  Would it work if we masked NaNs only (and left inf and -inf alone)?

**Comment by WeatherGod** [comment]
why not just use vmax + eps and vmin - eps? Makes for a very small chance
for overflow.

**Comment by mdboom** [comment]
@WeatherGod: That does indeed fix the test.  @efiring: Any downsides to that you can think of?

**Comment by efiring** [comment]
I thought of that also, but I wasn't sure whether vmin/vmax would always be known and invariant at that point in the execution.

**Comment by efiring** [comment]
Also, when you norm vmax + eps, is it guaranteed that the result will be greater than 1?  I doubt it.

File: lib/matplotlib/contour.py [hunk 1]
```diff
@@ -1231,10 +1231,11 @@ def _process_levels(self):
         self.layers = 0.5 * (self._levels[:-1] + self._levels[1:])
         # ...except that extended layers must be outside the
         # normed range:
+        finfo = np.finfo(float)
```
Comment (tacaswell) [inline lib/matplotlib/contour.py]: iirc, this can be a major time sink

Comment (mdboom) [inline lib/matplotlib/contour.py]: Yeah.  In this case, I don't think it's in a hot path, though.

**Comment by efiring** [comment]
@mdboom, Regarding an earlier question: I think that masking only the nan values and leaving inf alone would work in this case, but would not be the right thing for pcolormesh in general.  It would mean that autoscaling of the data range would still give useless results.  I think using `masked_invalid` is correct, and that something like my original suggestion, with genuine numbers, is the way to handle this.  It might not look pure, but it is practical.

**Comment by tacaswell** [comment]
@efiring It looks like this `self.vmin` is unrelated to the `self.vmin` in the `Normalize` classes as noted by you in 9cba6fb3a7aa83d5276eb so in this case we do know `vmin`/`vmax` as they are extracted from the levels.   

I think it is better to use the `eps` approach than a fixed large number to ensure that `self.layers` is always strictly increasing.

**Comment by tacaswell** [comment]
backported as d1f525cbb4fa3d27be0cc823eb9566b51076cab9

**Comment by efiring** [comment]
Why was this merged?  I don't think it is correct.  I don't see any reply to my objections above.

**Comment by efiring** [comment]
Confusing indeed.  Can we just delete self.vmin and self.vmax here?
In any case, even with local variables for the vmin and vmax, this still would not be not correct. The extend values need to be outside the *normed* range.  This changeset is just putting them marginally outside the *data* range.  @tacaswell, please revert this.

**Comment by tacaswell** [comment]
reverted on master as f90367a   reverted on v1.5.x as 605718b281c4cf9534271de064e5b5318d6a3ca1

Sorry, I miss understood.  I am not sure how to guarantee that the values normalize to outside the bounds for a general norm.   Maybe we should just use the top and bottom values in `self._levels` as if `extend` is true, they are already padded out.

**Comment by efiring** [comment]
I think the problem is that their padding is enough to bracket the "outside" data values, but not necessarily to ensure that the normed top and bottom levels are outside 0-1 range.  I stand by my earlier recommendation to use magic numbers which work (e.g. anything in the 1e100 to 1e300 range; let's split it and use 1e150) instead of magic numbers (e.g. finfo stuff) that *don't* work.

**Pull request merged**